tests/docs/test_complete_examples.py

```diff
@@ -148,11 +151,31 @@ def _get_definition_type(self):
                 self.source,
             )
         )
-        if len(definition_function_call_matches) != 1:
-            # Either no match,
-            # or we have both create_dataset() and create_measures().
-            # Both of these cases are invalid.
+        if len(definition_function_call_matches) > 1:
+            # we have both create_dataset() and create_measures().
             return None
+        if len(definition_function_call_matches) == 0:
+            # no create_dataset() or create_measures(); work out if it's a dataset definition
+            # file it is and add the relevant boilerplate
+            if "measure" in self.source:
+                # We're not currently handling incomplete measures examples
+                return None
+            else:
+                if "dataset.define_population" in self.source:
+                    population_definition = ""
+                else:
+                    population_definition = (
+                        "dataset.define_population(patients.exists_for_patient())"
+                    )
+                # Add imports and create_dataset/define_population at the beginning of the file
+                # to ensure it's available for the rest of the snippet. This results in a
+                # weird order of imports if the snippet already has imports, but it's still valid ehrql
+                self.source = (
+                    f"from ehrql import create_dataset\nfrom ehrql.tables.core import patients\n"
+                    f"dataset = create_dataset()\n{population_definition}\n"
+                    f"{self.source}"
```

Comment:
I've a mild preference for a multiline string with `textwrap.dedent()`.

Reply:
I tried dedent, but I think I'm doing it wrong

This doesn't indent a tutorial example properly:
```
self.source = textwrap.dedent(
    f"""from ehrql import create_dataset
    from ehrql.tables.core import patients
    dataset = create_dataset()
    {population_definition}
    {self.source}"""
)
```

An example from the tutorial that has the boilerplate added ends up indented like this:
```
from ehrql import create_dataset
                    from ehrql.tables.core import patients
                    dataset = create_dataset()
                    dataset.define_population(patients.exists_for_patient())
                    from ehrql import show
from ehrql.tables.core import patients, practice_registrations, clinical_events, medications

index_date = "2024-03-31"

show(
    patients.age_on(index_date),
    patients.is_alive_on(index_date)
)
```
I think the problem is that `self.source` has linebreaks in it, so `textwrap.dedent` doesn't recognise the leading whitespace as common to the other lines.

Happy to update if you know the way to make dedent do the right thing though

Reply:
The trick with `dedent` is a bit non-obvious but you need to suppress the leading newline with a `\`:
```py
self.source = textwrap.dedent(
    f"""\
    from ehrql import create_dataset
    from ehrql.tables.core import patients
    dataset = create_dataset()
    {population_definition}
    {self.source}"""
)
```

Because this is a bit non-obvious and easy to forget I also added a `strip_indent` function in the [ehrql.utils.string_utils](https://github.com/opensafely-core/ehrql/blob/4ea7bda492870bde3fa4794206109e4156d3f404/ehrql/utils/string_utils.py)
module, which you could use instead.

Reply:
That still doesn't work (although different output) - now I get:
```
                    from ehrql import create_dataset
                    from ehrql.tables.core import patients
                    dataset = create_dataset()
                    dataset.define_population(patients.exists_for_patient())
                    from ehrql import show
from ehrql.tables.core import patients, practice_registrations, clinical_events, medications

index_date = "2024-03-31"

show(
    patients.age_on(index_date),
    patients.is_alive_on(index_date)
)
```

And if I use `strip_indent`, I get the same output as before (without the leading \).  I still think it's because `self.source` contains `\n`s

Reply:
Yes, so if I remove all the `\n` from `self.source`, it's obviously broken, but it does dedent properly
```
source = self.source.replace("\n", "")
    self.source = strip_indent(
        f"""
        from ehrql import create_dataset
        from ehrql.tables.core import patients
        dataset = create_dataset()
        {population_definition}
        {source}
        """
    )
```
is formatted as 
```
from ehrql import create_dataset
from ehrql.tables.core import patients
dataset = create_dataset()
dataset.define_population(patients.exists_for_patient())
from ehrql import showfrom ehrql.tables.core import patients, practice_registrations, clinical_events, medicationsindex_date = "2024-03-31"show(    patients.age_on(index_date),    patients.is_alive_on(index_date))
```

Reply:
Given that @inglesp only expressed a mild preference for using dedent, I think I'll just leave it as it is

Reply:
Ah sorry, yeah! I wasn't paying enough attention here and saw one problem and assumed that was it.

I've been bitten by exactly this before, it's very annoying. Doing it the way you originally had it is probably the simplest solution. The other one is to use a template string rather than an f-string e.g.
```py
self.source = strip_indent(
    """
    from ehrql import create_dataset
    from ehrql.tables.core import patients
    dataset = create_dataset()
    {population_definition}
    {source}
    """
).format(
    source=self.source,
    population_definition=population_definition,
)
```

Reply:
Nope, template string still doesn't work (either with `string_indent` or `textwrap.dedent`); I'm getting the same indenting as before.

Reply:
I just tried the template thing locally and it did what I expected. You are removing the indent _before_ calling `.format()`, right?

Reply:
I swear I did, and I tried it both with `textwrap.dedent` (with the newline suppression) and with `strip_indent`, and it didn't work. But I just tried it again locally and now it does.